frame/contracts/src/wasm/runtime.rs

```diff
@@ -947,243 +948,292 @@ where
 // Any input that leads to a out of bound error (reading or writing) or failing to decode
 // data passed to the supervisor will lead to a trap. This is not documented explicitly
 // for every function.
-define_env!(Env, <E: Ext>,
-	// Account for used gas. Traps if gas used is greater than gas limit.
-	//
-	// NOTE: This is a implementation defined call and is NOT a part of the public API.
-	// This call is supposed to be called only by instrumentation injected code.
-	//
-	// - amount: How much gas is used.
-	[seal0] gas(ctx, amount: u32) => {
+#[define_env]
+pub mod env {
```

Comment:
Just an idea, because this is expanded as `pub struct Env`, perhaps we could instead annotate an `impl` block of a user defined `Env` struct instead of the module. That way all of the assumed in scope types e.g. `Runtime`, `TrapReason` etc will actually be in scope for that impl block

Reply:
I think it is a good idea. We could drop the module in that case because the other names the struct.

Reply:
Something to explore here might be to add the context to the struct e.g.

```
struct Env<'a, E: Ext> {
  ctx: &'a Runtime<E>,
}

// require impl From &Runtime<E> for macro to call

#[define_env]
impl<'a, E: Ext> {
  fn whatevs(&self, foo: u32) -> Result<(), TrapError> {
    self.ctx..whatevs(foo)
  }
}
```
That way it is not brittle to the naming of the `ctx` function parameter, which is an implicit requirement as the first parameter from what I can tell.

Reply:
This is a nice idea. However, I'm afraid the way of its implementation proposed is not realistic. Because of macro Hygiene rules, we can't do this in the way so that expanded code uses `&self` reference whilst the meant structure `Env` being defined outside the macro. How about leaving these experiments for a follow-up task, and keep the current implementation?